pkg/api/api.go

```diff
@@ -0,0 +1,233 @@
+
+	// path setup
+	router chi.Router
+	apis   []API
+
+	// authentication
+	keys map[string]struct{}
+
+	// info
+	logger  log.Logger
+	metrics bool
+
+	// throttling parameters
+	perMinRateLimit       int
+	burstLimit            int
+	concurrentAccessLimit int
+}
+
+// APIKeys allows for quick key validation
+type APIKeys map[string]struct{}
+
+// unique keys for context setting
+type key int
+
+const (
+	apiKeyCtxKey key = iota
+)
+
+// errors
+const (
+	errContextCanceled = "Context was canceled."
+)
+
+// Option allows to set optional parameters in the server
+type Option func(*Server)
+
+// WithKeys registers API authentication keys with the server
+func WithKeys(keys []string) Option {
+	return func(s *Server) {
+		for _, key := range keys {
+			s.keys[key] = struct{}{}
+		}
+	}
+}
+
+// WithLogger provides the api with access to the program level logger. It is recommended to use this option
+func WithLogger(l log.Logger) Option {
+	return func(s *Server) {
+		s.logger = l
+	}
+}
+
+// WithMetricsExport switches on metrics export
+func WithMetricsExport() Option {
+	return func(s *Server) {
+		s.metrics = true
+	}
+}
+
+// WithRateLimits applies custom request rate limits to the API. Setting any of the values to zero deactivates the rate limiting for that particular part.
+func WithRateLimits(perMin, burst, concurrentRequests int) Option {
+	return func(s *Server) {
+		s.perMinRateLimit = perMin
+		s.burstLimit = burst
+		s.concurrentAccessLimit = concurrentRequests
+	}
+}
+
+func getLoggerHandler(logger log.Logger) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			logger.Debugf("%s %s", r.Method, r.URL.Path)
+			next.ServeHTTP(w, r)
+		}
+		return http.HandlerFunc(fn)
+	}
+}
+
+func getMetrics(w http.ResponseWriter, r *http.Request) {
+	expvar.Handler().ServeHTTP(w, r)
+}
+
+// New creates a base router for goProbe's APIs and provides the metrics export out of the box
+func New(host, port string, manager *capture.Manager, opts ...Option) (*Server, error) {
+
+	s := &Server{
+		root:                  "/",
+		keys:                  make(map[string]struct{}),
+		perMinRateLimit:       defaultPerMinRateLimit,
+		burstLimit:            defaultBurstLimit,
+		concurrentAccessLimit: defaultConcurrentAccessLimit,
+		logger:                log.NewDevNullLogger(), // default is logging to nothing
+	}
+
+	if host == "" {
+		host = "localhost"
+	}
+	if port == "" {
+		return nil, fmt.Errorf("no port provided")
+	}
+	s.location = host + ":" + port
+
+	// apply options
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	// initialize currently supported APIs
+	s.apis = append(s.apis, v1.New(manager, v1.WithLogger(s.logger)))
+
+	r := chi.NewRouter()
+
+	// setup a good base middleware stack
+	r.Use(middleware.RequestID)
+	r.Use(middleware.RealIP)
+
+	// only use the logging middleware if a logger was specifically provided
+	switch s.logger.(type) {
```

Comment:
Couldn't you just use a simple type assertion instead of the rather clumsy switch statement, like:
`if _, isNullLogger := s.logger.(*log.DevNullLogger); !isNullLogger {
  r.use(reqLogger)
}`

Reply:
I could. Always found the switch statement less clumsy...